fd/diff.go

```diff
@@ -19,25 +18,30 @@ type Point struct {
 	Coeff float64
 }
 
-// Method represents a finite difference formula that approximates
-// the derivative of order k of a function f at x as
-//  d^k f(x) ≅ (1 / h^k) * \sum_i Coeff_i * f(x + h * Loc_i),
-// where h is a small positive step.
-type Method struct {
+// Formula represents a finite difference formula on a regularly spaced grid
+// that approximates the derivative of order k of a function f at x as
+//  d^k f(x) ≅ (1 / Step^k) * \sum_i Coeff_i * f(x + Step * Loc_i).
+type Formula struct {
 	// Stencil is the set of sampling Points which are used to estimate the
 	// derivative. The locations will be scaled by Step and are relative to x.
 	Stencil []Point
 	Order   int     // The order of the approximated derivative.
-	Step    float64 // Default step size for the method.
+	Step    float64 // Default step size for the formula.
```

Comment:
> Does that formula come from somewhere? It would never have occurred to me.

The optimal step size for forward difference for well-behaved functions is of the order of sqrt(eps). The default step sizes we have now try to keep the denominator in the finite difference formula at 1e-6. My formula is just an ad hoc way of combining the two things, and should be better ignored.

> It's reasonable to have a default I think.

Agreed. We can postpone the discussion on concrete values to another PR. We should also consider what the default method should be. Is the central approximation with its inevitable two evaluations really always necessary? Perhaps the forward approximation should be the default and where the extra accuracy is needed, the central approximation can be requested explicitly.